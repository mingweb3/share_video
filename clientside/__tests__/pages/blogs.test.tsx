/* eslint-disable import/no-extraneous-dependencies */
/**
 * @jest-environment jsdom
 */
import { render, screen } from '@testing-library/react'
import 'whatwg-fetch'
import { rest } from 'msw'
import { setupServer } from 'msw/node'

import BlogsPage from '@/app/blogs/page'

const apiBase = process.env.API_BASE
export const postsMock = [
  {
    userId: 1,
    id: 1,
    title: 'Title of the post ABC song',
    body: 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'
  },
  {
    userId: 1,
    id: 2,
    title: 'Title of the post ABC song 2',
    body: 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'
  },
  {
    userId: 1,
    id: 3,
    title: 'Title of the post ABC song 3',
    body: 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'
  }
]
// Setup REQUEST SERVER
const handlers = [
  rest.get(`${apiBase}/posts`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(postsMock))
  })
]
const server = setupServer(...handlers)

// Start request server
const consoleSpy = jest.spyOn(console, 'log').mockImplementation()
beforeAll(() => server.listen())
afterEach(() => {
  server.restoreHandlers()
  consoleSpy.mockClear()
})
afterAll(() => server.close())

// *** TESTING ***
describe('Render Blogs Page', () => {
  it('First Heading', async () => {
    const jsx = await BlogsPage()
    render(jsx)

    // Find Element to test
    const headingOne = screen.queryByRole('heading', { name: 'Blog' })

    expect(headingOne).toBeInTheDocument()
    expect(headingOne).toHaveClass('mb-2 text-2xl font-bold')
  })

  it('Description Heading', async () => {
    const jsx = await BlogsPage()
    const { container } = render(jsx)

    // Find Element to test
    // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
    const desc = container.querySelector('.page-desc')

    expect(desc).toBeInTheDocument()
    expect(desc).toHaveTextContent(/I will get Posts/i)
  })

  it('Blogs Page renders correctly with Snapshot', async () => {
    const jsx = await BlogsPage()
    const { container } = render(jsx)
    expect(container).toMatchSnapshot()
  })
})

describe('Render Post items', () => {
  it('Render enough posts', async () => {
    const jsx = await BlogsPage()
    const { container } = render(jsx)

    // Find Element to test
    // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
    const posts = container.querySelectorAll('.post-item')

    expect(posts.length).toBe(postsMock.length)
  })
})
