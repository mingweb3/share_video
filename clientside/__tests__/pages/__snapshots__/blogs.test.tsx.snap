// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Blogs Page Blogs Page renders correctly with Snapshot 1`] = `
<div>
  <div
    class="container mx-auto px-[20px] md:px-0"
  >
    <div
      class="inner py-6"
    >
      <h1
        class="mb-2 text-2xl font-bold"
      >
        Blog
      </h1>
      <div
        class="page-desc pt-8"
      >
        I will get Posts and show in a list (Nested Comp) - SSR
      </div>
      <div
        class="pt-8"
      >
        <div
          class="flex flex-col gap-6"
        >
          <div
            class="post-item rounded-lg border border-[#222222] p-4"
          >
            <h4
              class="mb-2 text-xl capitalize"
            >
              Title of the post ABC song
            </h4>
            <div
              class="mb-4 text-[#999999]"
            >
              quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto
            </div>
            <a
              class="italic text-[#dddddd] underline hover:text-white"
              href="/blogs/1"
            >
              Read more
            </a>
          </div>
          <div
            class="post-item rounded-lg border border-[#222222] p-4"
          >
            <h4
              class="mb-2 text-xl capitalize"
            >
              Title of the post ABC song 2
            </h4>
            <div
              class="mb-4 text-[#999999]"
            >
              quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto
            </div>
            <a
              class="italic text-[#dddddd] underline hover:text-white"
              href="/blogs/2"
            >
              Read more
            </a>
          </div>
          <div
            class="post-item rounded-lg border border-[#222222] p-4"
          >
            <h4
              class="mb-2 text-xl capitalize"
            >
              Title of the post ABC song 3
            </h4>
            <div
              class="mb-4 text-[#999999]"
            >
              quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto
            </div>
            <a
              class="italic text-[#dddddd] underline hover:text-white"
              href="/blogs/3"
            >
              Read more
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
